### ISSUE TITLE
[X86] 256-bit loads and stores for AVX512 vector types are suboptimal

### ISSUE THREAD

**Issue by ZviRackover** [halide/Halide#1644]
At the bottom there are the following:
1.  A modified version of lesson_11_cross_compilation.cpp with a AVX512 target and vectorization width of 16. 16 x uint32 is naturally mapped to ZMM registers.
2. The hot loop body's basic-block extracted from the LLVM IR dumped in debug mode.
3. The generated assembly code for 2.

Assuming that 1. is ok...
You can see in 2.  that although the vectorization width is 16, the loads and store are of type <8 x i32>.
The generated code in 3. shows one 512-bit load folded into the 'paddd' instruction, but there are two 256-bit stores. I have seen cases where both the loads and the stores are split, but this example is sufficient.

It would be more efficient for both loads and store to be 512-bits.
```
#include "Halide.h"
using namespace Halide;

int main(int argc, char **argv) {

    Func brighter;
    Var x, y;

    Param<uint32_t> offset;
    ImageParam input(type_of<uint32_t>(), 2);
    std::vector<Argument> args(2);
    args[0] = input;
    args[1] = offset;

    brighter(x, y) = input(x, y) + offset;
    brighter.vectorize(x, 16);

    Target target;
    target.os = Target::Linux;
    target.arch = Target::X86;
    target.bits = 64;
    std::vector<Target::Feature> x86_features;
    x86_features.push_back(Target::AVX512_Skylake);
    x86_features.push_back(Target::AVX2);
    x86_features.push_back(Target::AVX);
    x86_features.push_back(Target::SSE41);
    x86_features.push_back(Target::NoAsserts);
    x86_features.push_back(Target::NoBoundsQuery);
    x86_features.push_back(Target::NoRuntime);
    target.set_features(x86_features);
    brighter.compile_to_assembly("test.s", args, "brighter", target);
    return 0;
}
```

```
"for f0.s0.v0.v0.us.prol":                        ; preds = %"for f0.s0.v0.v0.us.prol.preheader"
  %43 = add i64 %41, %tmp37.us
  %44 = getelementptr inbounds i32, i32* %p1.host54, i64 %43
  %45 = bitcast i32* %44 to <8 x i32>*
  %46 = load <8 x i32>, <8 x i32>* %45, align 4, !tbaa !6
  %47 = add nsw i64 %43, 8
  %48 = getelementptr inbounds i32, i32* %p1.host54, i64 %47
  %49 = bitcast i32* %48 to <8 x i32>*
  %50 = load <8 x i32>, <8 x i32>* %49, align 4, !tbaa !6
  %51 = shufflevector <8 x i32> %46, <8 x i32> %50, <16 x i32> <i32 0, i32 1, i32 2, i32 3, i32 4, i32 5, i32 6, i32 7, i32 8, i32 9, i32 10, i32 11, i32 12, i32 13, i32 14, i32 15>
  %52 = add <16 x i32> %51, %17
  %53 = shufflevector <16 x i32> %52, <16 x i32> undef, <8 x i32> <i32 0, i32 1, i32 2, i32 3, i32 4, i32 5, i32 6, i32 7>
  %54 = getelementptr inbounds i32, i32* %f0.host55, i64 %tmp35.us
  %55 = bitcast i32* %54 to <8 x i32>*
  store <8 x i32> %53, <8 x i32>* %55, align 4, !tbaa !9
  %56 = shufflevector <16 x i32> %52, <16 x i32> undef, <8 x i32> <i32 8, i32 9, i32 10, i32 11, i32 12, i32 13, i32 14, i32 15>
  %57 = add nsw i64 %tmp35.us, 8
  %58 = getelementptr inbounds i32, i32* %f0.host55, i64 %57
  %59 = bitcast i32* %58 to <8 x i32>*
  store <8 x i32> %56, <8 x i32>* %59, align 4, !tbaa !9
  br label %"for f0.s0.v0.v0.us.prol.loopexit.unr-lcssa"

```

```
.LBB0_16:                               # %"for f0.s0.v0.v0.us.prol"
                                        #   in Loop: Header=BB0_14 Depth=1
	movq	-16(%rsp), %rcx         # 8-byte Reload
	leaq	(%rcx,%r10), %rdx
	vpaddd	(%r15,%rdx,4), %zmm0, %zmm1
	vmovdqu32	%ymm1, (%r12,%r13,4)
	vextracti32x8	$1, %zmm1, 32(%r12,%r13,4)
	movl	$1, %edi
.LBB0_17:                               # %"for f0.s0.v0.v0.us.prol.loopexit"

```

**Comment by abadams** [comment]
Looks like it's making the decision based on the AVX512 target feature, and
ignoring more specific Target features (AVX512_Skylake). Will fix.

**Comment by abadams** [comment]
Should be fixed by d81b6b44974ead9990a2c38b209769fe1fc04801

**Comment by ZviRackover** [comment]
Thanks!
